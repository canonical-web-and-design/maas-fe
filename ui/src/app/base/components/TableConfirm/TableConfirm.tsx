import type { ReactNode } from "react";

import {
  ActionButton,
  Button,
  Col,
  Notification,
  Row,
} from "@canonical/react-components";
import type { Props as ButtonProps } from "@canonical/react-components/dist/components/Button";

import { COL_SIZES } from "app/base/constants";
import { useCycled } from "app/base/hooks";

export type Props = {
  confirmAppearance?: ButtonProps["appearance"];
  confirmLabel: string;
  errors?: string | Record<string, string[]>;
  errorKey?: string;
  finished: boolean;
  inProgress: boolean;
  message: ReactNode;
  onClose: () => void;
  onConfirm: () => void;
  onSuccess?: () => void;
  sidebar?: boolean;
};

const TableConfirm = ({
  confirmAppearance = "positive",
  confirmLabel,
  errors,
  errorKey,
  finished,
  inProgress,
  message,
  onClose,
  onConfirm,
  onSuccess,
  sidebar = true,
}: Props): JSX.Element => {
  useCycled(finished, () => {
    onSuccess && onSuccess();
    onClose();
  });
  let errorMessage: string | null = null;
  if (errors) {
    if (typeof errors === "string") {
      errorMessage = errors;
    } else if (errorKey && typeof errors === "object") {
      errorMessage = errors[errorKey].join(" ");
    }
  }

  const { TABLE_CONFIRM_BUTTONS, SIDEBAR, TOTAL } = COL_SIZES;
  return (
    <Row>
      {errorMessage && (
        <Notification type="negative" status="Error:">
          {errorMessage}
        </Notification>
      )}
      <Col
        size={
          sidebar
            ? TOTAL - SIDEBAR - TABLE_CONFIRM_BUTTONS
            : TOTAL - TABLE_CONFIRM_BUTTONS
        }
      >
        <p className="u-no-margin--bottom u-no-max-width">{message}</p>
      </Col>
      <Col size={TABLE_CONFIRM_BUTTONS} className="u-align--right">
        <Button
          className="u-no-margin--bottom"
          data-test="action-cancel"
          onClick={onClose}
        >
          Cancel
        </Button>
        <ActionButton
          appearance={confirmAppearance}
          className="u-no-margin--bottom"
          data-test="action-confirm"
          loading={inProgress}
          onClick={onConfirm}
          success={finished}
          type="button"
        >
          {confirmLabel}
        </ActionButton>
      </Col>
    </Row>
  );
};

export default TableConfirm;
