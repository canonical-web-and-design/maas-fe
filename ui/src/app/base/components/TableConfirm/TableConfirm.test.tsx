import { mount, shallow } from "enzyme";

import TableConfirm from "./TableConfirm";

describe("TableConfirm", () => {
  it("renders", () => {
    const wrapper = shallow(
      <TableConfirm
        confirmLabel="save"
        finished={false}
        inProgress={false}
        message="Are you sure"
        onClose={jest.fn()}
        onConfirm={jest.fn()}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("can confirm", () => {
    const onConfirm = jest.fn();
    const wrapper = shallow(
      <TableConfirm
        confirmLabel="save"
        finished={false}
        inProgress={false}
        message="Are you sure"
        onClose={jest.fn()}
        onConfirm={onConfirm}
      />
    );
    wrapper.find("ActionButton").simulate("click");
    expect(onConfirm).toHaveBeenCalled();
  });

  it("can cancel", () => {
    const onClose = jest.fn();
    const wrapper = shallow(
      <TableConfirm
        confirmLabel="save"
        finished={false}
        inProgress={false}
        message="Are you sure"
        onClose={onClose}
        onConfirm={jest.fn()}
      />
    );
    wrapper.find("Button").simulate("click");
    expect(onClose).toHaveBeenCalled();
  });

  it("closes when it has finished", () => {
    const onClose = jest.fn();
    const wrapper = mount(
      <TableConfirm
        confirmLabel="save"
        finished={false}
        inProgress={false}
        message="Are you sure"
        onClose={onClose}
        onConfirm={jest.fn()}
      />
    );
    wrapper.find("ActionButton").simulate("click");
    expect(onClose).not.toHaveBeenCalled();
    wrapper.setProps({ finished: true });
    wrapper.update();
    expect(onClose).toHaveBeenCalled();
  });

  it("runs onSuccess function when it has finished", () => {
    const onSuccess = jest.fn();
    const wrapper = mount(
      <TableConfirm
        confirmLabel="save"
        finished={false}
        inProgress={false}
        message="Are you sure"
        onClose={jest.fn()}
        onConfirm={jest.fn()}
        onSuccess={onSuccess}
      />
    );
    wrapper.find("ActionButton").simulate("click");
    expect(onSuccess).not.toHaveBeenCalled();
    wrapper.setProps({ finished: true });
    wrapper.update();
    expect(onSuccess).toHaveBeenCalled();
  });

  it("can display an error", () => {
    const onClose = jest.fn();
    const wrapper = mount(
      <TableConfirm
        confirmLabel="save"
        errors="It didn't work"
        finished={false}
        inProgress={false}
        message="Are you sure"
        onClose={onClose}
        onConfirm={jest.fn()}
      />
    );
    expect(wrapper.find("Notification").text()).toBe("Error:It didn't work");
  });

  it("can display an error for a field", () => {
    const onClose = jest.fn();
    const wrapper = mount(
      <TableConfirm
        confirmLabel="save"
        errors={{ delete: ["It didn't work"] }}
        errorKey="delete"
        finished={false}
        inProgress={false}
        message="Are you sure"
        onClose={onClose}
        onConfirm={jest.fn()}
      />
    );
    expect(wrapper.find("Notification").text()).toBe("Error:It didn't work");
  });
});
