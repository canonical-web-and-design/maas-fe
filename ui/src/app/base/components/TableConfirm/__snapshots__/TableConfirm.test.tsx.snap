// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableConfirm renders 1`] = `
<Row>
  <Col
    size={5}
  >
    <p
      className="u-no-margin--bottom u-no-max-width"
    >
      Are you sure
    </p>
  </Col>
  <Col
    className="u-align--right"
    size={4}
  >
    <Button
      className="u-no-margin--bottom"
      data-test="action-cancel"
      onClick={[MockFunction]}
    >
      Cancel
    </Button>
    <ActionButton
      appearance="positive"
      className="u-no-margin--bottom"
      data-test="action-confirm"
      loading={false}
      onClick={[MockFunction]}
      success={false}
      type="button"
    >
      save
    </ActionButton>
  </Col>
</Row>
`;
